data-processor/src/main/java/io/micronaut/data/processor/visitors/finders/AbstractCriteriaMethodMatch.java

```diff
@@ -825,8 +826,13 @@ protected List<SourcePersistentProperty> getDtoProjectionProperties(SourcePersis
                     // Convert anything to a string or an object
                     return pp;
                 }
-                if (!TypeUtils.areTypesCompatible(dtoPropertyType, pp.getType())) {
-                    throw new MatchFailedException("Property [" + propertyName + "] of type [" + dtoPropertyType.getName() + "] is not compatible with equivalent property of type [" + pp.getType().getName() + "] declared in entity: " + entity.getName());
+                boolean compatibleTypes = TypeUtils.areTypesCompatible(dtoPropertyType, pp.getType());
+                if (!compatibleTypes) {
```

Comment:
This might be the fix for initial issue. If field types are not simple field or same classes, then check if entity matches with DTO.
After this, new error is when reading data 
> io.micronaut.data.exceptions.DataAccessException: Error reading object for name [author] from result set: Column "author" not found [42122-224]

because mapper tries to read it into the field vs into joined object. Also, should method specify join type for the DTO or automatically assume join is requested?